lib/util/schema.ts

```diff
@@ -0,0 +1,59 @@
+import hasha from 'hasha';
+import type { z } from 'zod';
+import { logger } from '../logger';
+import * as memCache from './cache/memory';
+import { safeStringify } from './stringify';
+
+type SchemaErrorsMap = Record<string, Record<string, z.ZodError>>;
+
+function getCacheKey(error: z.ZodError): string {
+  const content = safeStringify(error);
+  const key = hasha(content).slice(0, 32);
+  return `err_${key}`;
+}
+
+function collectError<T extends z.ZodSchema>(
+  schema: T,
+  error: z.ZodError
+): void {
+  const { description = 'Unspecified schema' } = schema;
+  const schemaErrorsMap = memCache.get<SchemaErrorsMap>('schema-errors') ?? {};
+  const schemaErrors = schemaErrorsMap[description] ?? {};
+  const key = getCacheKey(error);
+  const schemaError = schemaErrors[key];
+  if (!schemaError) {
+    schemaErrors[key] = error;
+    schemaErrorsMap[description] = schemaErrors;
+  }
+  memCache.set('schema-errors', schemaErrorsMap);
+}
+
+export function reportErrors(): void {
+  const schemaErrorsMap = memCache.get<SchemaErrorsMap>('schema-errors');
+  if (!schemaErrorsMap) {
+    return;
+  }
+
+  for (const [description, schemaErrors] of Object.entries(schemaErrorsMap)) {
+    const errors = Object.values(schemaErrors);
+    for (const err of errors) {
+      logger.warn({ description, err }, `Schema validation error`);
+    }
+  }
+
+  memCache.set('schema-errors', null);
+}
+
+export function checkSchema<T extends z.ZodSchema>(
+  schema: T,
+  input: unknown
+): input is z.infer<T> {
+  const res = schema.safeParse(input);
+  const { success } = res;
+  if (!success) {
+    collectError(schema, res.error);
```

Comment:
why do we defer error reporting? isn't it hard to find right code later then?

Reply:
Can be found by stacktrace. For now, it's more like additional stats reporting rather than runtime errors.

Reply:
Oh yes, we do, `collectError` handles it